modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/SqlRowHandler.java

```diff
@@ -0,0 +1,272 @@
+        List<TypeSpec> rightTypes = right.rowSchema().fields();
+
+        Object[] values = new Object[leftLen + rightLen];
+        Builder schemaBuilder = RowSchema.builder();
+
+        for (int i = 0; i < leftLen; i++) {
+            values[i] = left.get(i);
+            schemaBuilder.addField(leftTypes.get(i));
+        }
+
+        for (int i = 0; i < rightLen; i++) {
+            values[leftLen + i] = right.get(i);
+            schemaBuilder.addField(rightTypes.get(i));
+        }
+
+        return new ObjectsArrayRowWrapper(schemaBuilder.build(), values);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public RowWrapper map(RowWrapper row, int[] mapping, int offset) {
+        Object[] fields = new Object[mapping.length];
+
+        for (int i = 0; i < mapping.length; i++) {
+            fields[i] = row.get(mapping[i] + offset);
+        }
+
+        return new ObjectsArrayRowWrapper(row.rowSchema(), fields);
+    }
+
+    @Override
+    public int columnCount(RowWrapper row) {
+        return row.columnsCount();
+    }
+
+    @Override
+    public String toString(RowWrapper row) {
+        IgniteStringBuilder buf = new IgniteStringBuilder("Row[");
+        int maxIdx = columnCount(row) - 1;
+
+        for (int i = 0; i <= maxIdx; i++) {
+            buf.app(row.get(i));
+
+            if (i != maxIdx) {
+                buf.app(", ");
+            }
+        }
+
+        return buf.app(']').toString();
+    }
+
+    @Override
+    public ByteBuffer toByteBuffer(RowWrapper row) {
+        return row.toByteBuffer();
+    }
+
+    @Override
+    public RowFactory<RowWrapper> factory(RowSchema rowSchema) {
+        int rowLen = rowSchema.fields().size();
+
+        return new RowFactory<>() {
+            /** {@inheritDoc} */
+            @Override
+            public RowHandler<RowWrapper> handler() {
+                return SqlRowHandler.this;
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper create() {
+                return create(new Object[rowLen]);
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper create(Object... fields) {
+                assert fields.length == rowLen;
+
+                return new ObjectsArrayRowWrapper(rowSchema, fields);
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper create(ByteBuffer buf) {
+                return wrap(new BinaryTuple(rowLen, buf));
+            }
+
+            /** {@inheritDoc} */
+            @Override
+            public RowWrapper wrap(InternalTuple tuple) {
+                assert rowLen == tuple.elementCount();
+
+                return new BinaryTupleRowWrapper(rowSchema, tuple);
+            }
+        };
+    }
+
+    /**
+     * Provides the ability for a single {@link RowHandler} instance to interact with multiple row implementations.
+     */
+    public abstract static class RowWrapper {
```

Comment:
i see an interface, not abstract class here )

Reply:
I would like to hide the methods of this class so that it is possible to work only through the RowHandler and not directly :thinking: :roll_eyes: